File: frontend/lib/models/service_history_part_dto.dart
class ServiceHistoryPartDto {
  final int? id;
  final int? serviceHistoryId;
  final int? partCatalogId;
  final String partName;
  final String catalogNumber;
  final int quantity;
  final double unitCost;
  final double? totalCost;
  final int? warrantyMonths;
  final int? supplierId;
  final String? supplierName;
  final String? installationNotes;
  final DateTime? createdDate;

  ServiceHistoryPartDto({
    this.id,
    this.serviceHistoryId,
    this.partCatalogId,
    required this.partName,
    required this.catalogNumber,
    required this.quantity,
    required this.unitCost,
    this.totalCost,
    this.warrantyMonths,
    this.supplierId,
    this.supplierName,
    this.installationNotes,
    this.createdDate,
  });

  factory ServiceHistoryPartDto.fromJson(Map<String, dynamic> json) {
    DateTime? d(dynamic v) => (v is String && v.isNotEmpty) ? DateTime.parse(v) : null;
    double? f(dynamic v) => v == null ? null : (v as num).toDouble();
    return ServiceHistoryPartDto(
      id: (json['id'] as num?)?.toInt(),
      serviceHistoryId: (json['serviceHistoryId'] as num?)?.toInt(),
      partCatalogId: (json['partCatalogId'] as num?)?.toInt(),
      partName: json['partName'] as String,
      catalogNumber: json['catalogNumber'] as String,
      quantity: (json['quantity'] as num).toInt(),
      unitCost: f(json['unitCost']) ?? 0.0,
      totalCost: f(json['totalCost']),
      warrantyMonths: (json['warrantyMonths'] as num?)?.toInt(),
      supplierId: (json['supplierId'] as num?)?.toInt(),
      supplierName: json['supplierName'] as String?,
      installationNotes: json['installationNotes'] as String?,
      createdDate: d(json['createdDate']),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'serviceHistoryId': serviceHistoryId,
        'partCatalogId': partCatalogId,
        'partName': partName,
        'catalogNumber': catalogNumber,
        'quantity': quantity,
        'unitCost': unitCost,
        'totalCost': totalCost,
        'warrantyMonths': warrantyMonths,
        'supplierId': supplierId,
        'supplierName': supplierName,
        'installationNotes': installationNotes,
        'createdDate': createdDate?.toIso8601String(),
      };
}

File: backend/src/main/java/ru/bowling/bowlingapp/DTO/ServiceHistoryPartDTO.java
package ru.bowling.bowlingapp.DTO;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import jakarta.validation.constraints.*;
import java.time.LocalDateTime;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class ServiceHistoryPartDTO {

    private Long id;
    private Long serviceHistoryId;
    private Long partCatalogId;

    @NotBlank(message = "Part name is required")
    @Size(max = 255, message = "Part name cannot exceed 255 characters")
    private String partName;

    @NotBlank(message = "Catalog number is required")
    @Size(max = 100, message = "Catalog number cannot exceed 100 characters")
    private String catalogNumber;

    @NotNull(message = "Quantity is required")
    @Min(value = 1, message = "Quantity must be at least 1")
    private Integer quantity;

    @NotNull(message = "Unit cost is required")
    @DecimalMin(value = "0.0", message = "Unit cost must be non-negative")
    private Double unitCost;

    private Double totalCost;

    @Min(value = 0, message = "Warranty period cannot be negative")
    private Integer warrantyMonths;

    private Long supplierId;
    private String supplierName;

    @Size(max = 1000, message = "Installation notes cannot exceed 1000 characters")
    private String installationNotes;

    private LocalDateTime createdDate;
}
